### PR TITLE
some gpio driver enhancements

### PR DESCRIPTION
**Pull request by vchong**
\+ @daniel-thompson fyi only

### CODE REVIEW

**Comment by d3zd3z** [comment]
I *think* these read8/write8 operations are going to be safe with regards to memory ordering. The kernel goes through tremendous effort to make sure there are iormb instructions after the operations, and such. I believe the IO operations themselves will be ordered properly, but it is possible they will not be ordered with regards to regular memory accesses. I think this driver is OK, though.

**Comment by vchong** [comment]
@d3zd3z Can you add R-b tag if ok?
Anyone else available for reviews?

File: core/drivers/pl061_gpio.c [hunk 1]
```diff
@@ -156,6 +191,8 @@ static const struct gpio_ops pl061_ops = {
 	.set_direction	= pl061_set_direction,
 	.get_value	= pl061_get_value,
 	.set_value	= pl061_set_value,
+	.get_interrupt = pl061_get_interrupt,
```
Comment (d3zd3z) [inline core/drivers/pl061_gpio.c]: The '=' don't quite match the above lines. The previous lines use a tab, and this is a single space. Suggest using a tab for consistency, even though it really shouldn't be a tab.

File: core/drivers/sub.mk [hunk 2]
```diff
@@ -1,15 +1,6 @@
...
-endif
-
-ifeq ($(CFG_PL022),y)
-$(call force,CFG_GPIO,y,required by CFG_PL022)
-endif
-
 srcs-$(CFG_CDNS_UART) += cdns_uart.c
 srcs-$(CFG_PL011) += pl011.c
 srcs-$(CFG_GIC) += gic.c
-srcs-$(CFG_GPIO) += gpio.c
```
Comment (d3zd3z) [inline core/drivers/sub.mk]: Should this patch also remove `gpio.c`?

**Comment by d3zd3z** [comment]
Added a couple of minor comments.

File: core/drivers/pl061_gpio.c (hunk 1 above)
Comment (vchong) [inline core/drivers/pl061_gpio.c]: I'll use a single space all the other instead. Thanks!

File: core/drivers/sub.mk (hunk 2 above)
Comment (vchong) [inline core/drivers/sub.mk]: Yes! Thanks for catching that!

**Comment by vchong** [comment]
Addressed review comments.

**Comment by d3zd3z** [comment]
```
Reviewed-by: David Brown <david.brown@linaro.org>
```

File: core/drivers/pl061_gpio.c [hunk 3]
```diff
@@ -134,24 +138,106 @@ static void pl061_set_value(unsigned int gpio_pin, enum gpio_level value)
...
+		data = read8(base_addr + GPIOIE) & ~BIT(offset);
+		write8(data, base_addr + GPIOIE);
+	}
+}
 
 /*
  * Register the PL061 GPIO controller with a base address and the offset
  * of start pin in this GPIO controller.
- * This function is called after pl061_gpio_ops_init().
+ * This function is called after pl061_gpio_init().
```
Comment (jforissier) [inline core/drivers/pl061_gpio.c]: `pl061_init()`?

Comment (vchong) [inline core/drivers/pl061_gpio.c]: Yes thanks! Forgot to update the comments.

**Comment by jforissier** [comment]
Looks OK.
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by vchong** [comment]
Rebased, squashed and added tags. Ok to merge. Thanks!

**Pull request merged**